README
Acme-CPANAuthors

INSTALLATION

To install this module, run the following commands:
    perl Makefile.PL
    make
    make test
    make install


COPYRIGHT AND LICENCE

Copyright (C) 2007 Kenichi Ishigaki

This program is free software; you can redistribute it and/or
modify it under the same terms as Perl itself.


Changes
Revision history for Acme-CPANAuthors

0.16 2011/03/09
  - listed several new modules: DualLife, GitHub, Search
  - ignore Search while looking for authors

0.15 2011/03/05
  - tweaked end of line handling
  - better API server error handling (RT #66394, thanks to ANDK)

0.14 2010/09/30
  - stopped trying hard to find cpan mirror in the hard drive,
    and try downloading index files if necessary.
    Set ACME_CPANAUTHORS_HOME environmental variable to use local
    indices.

0.13 2010/09/26
  - forgot to remove Parse::CPAN::* bit in a test

0.12 2010/09/25
  - added ToBeLike to the list.
  - bundled Parse::CPAN::* alternatives to cut off antlers.

0.11 2010/09/20
  - oops, forgot to update dependencies

0.10 2010/09/20
  - use Try::Tiny for better trapping
  - added several modules to the subordinate modules list
  - launched a new api server to fetch kwalitee data as json.
    no need to scrape kwalitee pages anymore

0.09 2009/08/23
  - added Arabic and German to the subordinate modules list,
    and rearranged the list a bit.
  - added a reason why I wrote this
  - added a tentative workaround to t/03_extra.t. hope this works...

0.08 2009/04/21
  - added more lists to the pod
  - added $VERSION to the ::Utils not to break installers (RT #43388)

0.07 2009/02/19
  - added 'look_for' method and its test
  - added a list of Acme::CPANAuthors:: modules to the pod

0.06 2008/11/01
  - removed CPAN::Config dependency (RT #40560 by Slaven Rezic)
  - try to find CPAN sources via CPANPLUS configuration too

0.05 2008/10/27
  - skip extra gravatar test if CPAN.pm is not properly configured

0.04 2008/07/14
  - bumped up the version req of Parse::CPAN::Authors to 2.27
  - try harder to find 01mailrc file when there's a local CPAN
    mirror in the urllist of CPAN::Config

0.03 2008/05/30
  - "What's actually broken is Parse::CPAN::Authors 2.26" release
  - tentatively skip a test if Parse::CPAN::Authors 2.26 is loaded.

0.02 2008/05/22
  - "YAPC::Asia 2008 hackathon" release
  - made Acme::CPANAuthors::(Specific)->authors context sensitive.
  - better cpan config loading

0.01 2007/12/26
  - ported from former Acme::JapaneseCPANAuthors (not released)


t/lib/Acme/CPANAuthors/Test.pm
package Acme::CPANAuthors::Test;

use strict;
use warnings;
use Acme::CPANAuthors::Register (
  ISHIGAKI => 'Kenichi Ishigaki'
);

1;


Makefile.PL
use strict;
use warnings;
use ExtUtils::MakeMaker;

WriteMakefile(
  NAME           => 'Acme::CPANAuthors',
  AUTHOR         => 'Kenichi Ishigaki <ishigaki@cpan.org>',
  VERSION_FROM   => 'lib/Acme/CPANAuthors.pm',
  ABSTRACT_FROM  => 'lib/Acme/CPANAuthors.pm',
  PREREQ_PM => {
    'Carp'                   => 0,
    'Class::Accessor::Fast'  => 0,
    'CPAN::DistnameInfo'     => 0,
    'Exporter'               => 0,
    'File::Spec'             => 0,
    'Gravatar::URL'          => 0,
    'IO::File'               => 0,
    'IO::Uncompress::Gunzip' => '2.008', # core in 5.10
    'JSON'                   => 0,
    'LWP::UserAgent'         => 0,
    'Module::Find'           => 0,
    'Test::More'             => '0.47',
    'Test::UseAllModules'    => '0.09',
    'Try::Tiny'              => 0,
    'version'                => '0.74',  # core in 5.10
  },
  ($ExtUtils::MakeMaker::VERSION >= 6.31
    ? (LICENSE => 'perl')
    : ()
  ),
);


lib/Acme/CPANAuthors/Utils.pm
package Acme::CPANAuthors::Utils;

use strict;
use warnings;
use Carp;
use base qw( Exporter );
use File::Spec;

our $VERSION   = '0.14'; # see RT #43388
our @EXPORT_OK = qw( cpan_authors cpan_packages );

my $CPANFiles = {};

sub clear_cached_cpan_files () { $CPANFiles = {}; }

sub cpan_authors () {
  unless ( $CPANFiles->{authors} ) {
    require Acme::CPANAuthors::Utils::Authors;
    $CPANFiles->{authors} =
      Acme::CPANAuthors::Utils::Authors->new( _cpan_authors_file() );
  }
  return $CPANFiles->{authors};
}

sub cpan_packages () {
  unless ( $CPANFiles->{packages} ) {
    require Acme::CPANAuthors::Utils::Packages;
    $CPANFiles->{packages} =
      Acme::CPANAuthors::Utils::Packages->new( _cpan_packages_file() );
  }
  return $CPANFiles->{packages};
}

sub _cpan_authors_file () {
  _cpan_file( authors => '01mailrc.txt.gz' );
}

sub _cpan_packages_file () {
  _cpan_file( modules => '02packages.details.txt.gz' );
}

sub _cpan_file {
  my ($dir, $basename) = @_;

  my $file;
  if ($ENV{ACME_CPANAUTHORS_HOME}) {
    $file = _catfile($ENV{ACME_CPANAUTHORS_HOME}, $dir, $basename);
    return $file if $file && -r $file;
  }
  require File::Path;
  for my $parent (File::Spec->tmpdir, '.') {
    my $tmpdir = File::Spec->catdir($parent, '.acmecpanauthors', $dir);
    eval { File::Path::mkpath($tmpdir) };
    next unless -d $tmpdir && -r _;
    $file = _catfile($tmpdir, $basename);
    my $how_old = -M $file;
    if (!-r $file or !$how_old or $how_old > 1) {
      require LWP::UserAgent;
      my $ua = LWP::UserAgent->new(env_proxy => 1);
      my $res = $ua->mirror('http://www.cpan.org/'.$dir.'/'.$basename, $file);
      next if $res->is_error;
    }
    return $file if -r $file;
  }
  croak "$basename not found";
}

sub _catfile { File::Spec->canonpath( File::Spec->catfile( @_ ) ); }

1;

__END__

=head1 NAME

Acme::CPANAuthors::Utils

=head1 DESCRIPTION

This may export several utility functions to use internally.

=head1 FUNCTIONS

=head2 cpan_authors (exportable)

returns a (probably cached) Parse::CPAN::Authors object.

=head2 cpan_packages (exportable)

returns a (probably cached) Parse::CPAN::Packages object.

=head2 clear_cached_cpan_files

clears cached Parse::CPAN::Authors/Packages objects.

=head1 AUTHOR

Kenichi Ishigaki, E<lt>ishigaki at cpan.orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2007 by Kenichi Ishigaki.

This program is free software; you can redistribute it and/or
modify it under the same terms as Perl itself.

=cut


lib/Acme/CPANAuthors/Register.pm
package Acme::CPANAuthors::Register;

use strict;
use warnings;

sub import {
  my ($class, %authors) = @_;

  my $caller = caller;
  {
    no strict 'refs';
    *{"$caller\::authors"} = sub { wantarray ? %authors : \%authors };
  }
}

1;

__END__

=head1 NAME

Acme::CPANAuthors::Register

=head1 SYNOPSIS

  package Acme::CPANAuthors::YourGroup
  use strict;
  use warnings;
  our $VERSION = '0.071226';
  use Acme::CPANAuthors::Register (
    ID  => 'Real Name',
  );

  1;

  # then you can get authors list like these.
  # note that ->authors is context sensitive)

  my %hash    = Acme::CPANAuthors::YourGroup->authors;
  my $hashref = Acme::CPANAuthors::YourGroup->authors;

=head1 DESCRIPTION

This is used to register Pause IDs and real names of those who belong to your country/local perl mongers group/your company etc.

=head1 SEE ALSO

L<Acme::CPANAuthors>

=head1 AUTHOR

Kenichi Ishigaki, E<lt>ishigaki at cpan.orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2007 by Kenichi Ishigaki.

This program is free software; you can redistribute it and/or
modify it under the same terms as Perl itself.

=cut


lib/Acme/CPANAuthors/Utils/Authors.pm
package Acme::CPANAuthors::Utils::Authors;

use strict;
use warnings;
use base 'Acme::CPANAuthors::Utils::CPANIndex';

sub _mappings {+{author => 'authors'}}

sub _parse {
  my ($self, $file) = @_;

  my $handle = $self->_handle($file);

  while (my $line = $handle->getline) {
    $line =~ s/\r?\n$//;
    my ($alias, $pauseid, $long) = split ' ', $line, 3;
    $long =~ s/^"//;
    $long =~ s/"$//;
    my ($name, $email) = $long =~ /(.*) <(.+)>$/;
    my $author = Acme::CPANAuthors::Utils::Authors::Author->new({
      pauseid => $pauseid,
      name    => $name,
      email   => $email,
    });

    $self->{authors}{$pauseid} = $author;
  }
}

package #
  Acme::CPANAuthors::Utils::Authors::Author;

use strict;
use warnings;
use base 'Class::Accessor::Fast';

__PACKAGE__->mk_ro_accessors(qw/pauseid name email/);

1;

__END__

=head1 NAME

Acme::CPANAuthors::Utils::Authors

=head1 SYNOPSIS

  use Acme::CPANAuthors::Utils::Authors;

  # you can't pass the raw content of 01mailrc.txt(.gz)
  my $authors = Acme::CPANAuthors::Utils::Authors->new(
    'cpan/authors/01mailrc.txt.gz'
  );

  my $author = $authors->author('Acme::CPANAuthors');

=head1 DESCRIPTION

This is a subset of L<Parse::CPAN::Authors>. The reading
methods are similar in general (accessors are marked as
read-only, though). Internals and data-parsing methods may
be different, but you usually don't need to care.

=head1 METHODS

=head2 new

always takes a file name (both raw C<.txt> file and C<.txt.gz>
file name are acceptable). Raw content of the file is not
acceptable.

=head2 author

takes a name of an author, and returns an object that represents it.

=head2 authors

returns a list of stored author objects.

=head2 author_count

returns the number of stored authors.

=head1 AUTHOR ACCESSORS

=head2 pauseid, name, email

  my $author = $authors->author('ISHIGAKI');
  print $author->pauseid, "\n"; # ISHIGAKI
  print $author->name, "\n";    # Kenichi Ishigaki
  print $author->email, "\n";   # ishigaki@cpan.org

=head1 SEE ALSO

L<Parse::CPAN::Authors>

=head1 AUTHOR

Kenichi Ishigaki, E<lt>ishigaki@cpan.orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2010 by Kenichi Ishigaki.

This program is free software; you can redistribute it and/or
modify it under the same terms as Perl itself.

=cut


lib/Acme/CPANAuthors/Utils/Kwalitee.pm
package Acme::CPANAuthors::Utils::Kwalitee;

use strict;
use warnings;
use LWP::UserAgent;
use JSON ();

my $ua;

sub _uri { "http://api.cpanauthors.org/kwalitee/" . shift }

sub _ua {
  my $class = shift;
  $ua = $_[0] if @_;
  $ua ||= LWP::UserAgent->new(env_proxy => 1);
  $ua;
}

sub fetch {
  my ($class, $id) = @_;

  return unless $id;

  my $res = $class->_ua->get(_uri(lc $id));
  unless ($res->is_success && $res->code == 200) {
    $class->_error($res->status_line);
  }

  my $json = eval { JSON::decode_json($res->content) };
  if ($@) {
    $class->_error($@);
  }
  return $json;
}

sub _error {
  my ($class, $error) = @_;

  die "API SERVER ERROR\n"
    . "Couldn't parse kwalitee info from the api server\n"
    . "  $error\n"
    . "Sorry for the inconvenience. If this lingers long,\n"
    . "please drop a line to <ishigaki\@cpan.org>.\n";
}

1;

__END__

=head1 NAME

Acme::CPANAuthors::Utils::Kwalitee

=head1 SYNOPSIS

    use Acme::CPANAuthors::Utils::Kwalitee;

    my $info = Acme::CPANAuthors::Utils::Kwalitee->fetch('ISHIGAKI');

    # then, $info will have keys/values like this
    # (though this may change without notices)

    {
      info => {
        Average_Kwalitee => "144.333",
        CPANTS_Game_Kwalitee => "98.833",
        Email => "ishigaki\@cpan.org",
        Liga => "5 or more",
        Rank => 47,
      },
      distributions => {
        "Acme-CPANAuthors" => {
          details  => { "use_strict" => "ok", ... },
          kwalitee => "144.000",
        },
        ...
      },
    }

=head1 DESCRIPTION

This is mainly to create some local kwalitee rankings for you :)

=head1 METHODS

=head2 fetch

returns kwalitee information for the author of the id.
This information is scraped from http://cpants.perl.org/.

=head1 AUTHOR

Kenichi Ishigaki, E<lt>ishigaki at cpan.orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2007 by Kenichi Ishigaki.

This program is free software; you can redistribute it and/or
modify it under the same terms as Perl itself.

=cut


lib/Acme/CPANAuthors/Utils/Packages.pm
package Acme::CPANAuthors::Utils::Packages;

use strict;
use warnings;
use CPAN::DistnameInfo;
use version;
use base 'Acme::CPANAuthors::Utils::CPANIndex';

sub _preambles {qw(
  file url description columns intended_for
  written_by line_count last_updated
)}

sub _mappings {+{
  package             => 'packages',
  distribution        => 'dists',
  latest_distribution => 'latest_dists',
}}

sub _parse {
  my ($self, $file) = @_;

  my $handle = $self->_handle($file);

  my $done_preambles = 0;
  while (my $line = $handle->getline) {
    $line =~ s/\r?\n$//;
    unless ($done_preambles) {
      if ($line =~ /^\s*$/) {
        $done_preambles = 1;
      }
      elsif (my ($key, $value) = $line =~ /^([^:]+):\s*(.*)/) {
        $key =~ tr/A-Z\-/a-z_/;
        $self->{preambles}{$key} = $value;
      }
      next;
    }

    my ($package, $version, $path) = split ' ', $line;

    my $dist = $self->_dist_from_path($path);

    my $pkg = Acme::CPANAuthors::Utils::Packages::Package->new({
      package      => $package,
      version      => $version,
      distribution => $dist,
    });

    push @{ $dist->{packages} ||= [] }, $pkg;

    $self->{packages}{$package} = $pkg;
  }
}

sub _dist_from_path {
  my ($self, $path) = @_;

  my $dist = $self->{dists}{$path};
  return $dist if $dist;

  my $info = CPAN::DistnameInfo->new($path);
  $dist = Acme::CPANAuthors::Utils::Packages::Distribution->new({
    prefix    => $path,
    dist      => $info->dist,
    version   => $info->version,
    maturity  => $info->maturity,
    filename  => $info->filename,
    cpanid    => $info->cpanid,
    distvname => $info->distvname,
  });

  $self->{dists}{$path} = $dist;

  return unless defined $dist->version && $dist->dist;

  # see if it's latest
  my $distname = $info->dist;
  my $latest = $self->{latest_dists}{$distname};
  unless ($latest) {
    $self->{latest_dists}{$distname} = $dist;
    return $dist;
  }
  my ($distv, $latestv);
  eval {
    no warnings;
    $distv   = version->new( $dist->version   || 0 );
    $latestv = version->new( $latest->version || 0 );
  };
  if ($distv && $latestv) {
    if ($distv > $latestv) {
      $self->{latest_dists}{$distname} = $dist;
    }
  }
  else {
    no warnings;
    if ($dist->version > $latest->version) {
      $self->{latest_dists}{$distname} = $dist;
    }
  }

  $dist;
}

package #
  Acme::CPANAuthors::Utils::Packages::Distribution;

use strict;
use warnings;
use base 'Class::Accessor::Fast';

__PACKAGE__->mk_ro_accessors(qw/
  prefix
  dist
  version
  maturity
  filename
  cpanid
  distvname
  packages
/);

*path    = *prefix;
*name    = *dist;
*pauseid = *cpanid;

sub contains { @{ shift->{packages} || [] } }

package #
  Acme::CPANAuthors::Utils::Packages::Package;

use strict;
use warnings;
use base 'Class::Accessor::Fast';

__PACKAGE__->mk_ro_accessors(qw/
  package
  version
  distribution
/);

*name = *package;

1;

__END__

=head1 NAME

Acme::CPANAuthors::Utils::Packages

=head1 SYNOPSIS

  use Acme::CPANAuthors::Utils::Packages;

  # you can't pass the raw content of 02packages.details.txt(.gz)
  my $packages = Acme::CPANAuthors::Utils::Packages->new(
    'cpan/modules/02packages.details.txt.gz'
  );

  my $package = $packages->package('Acme::CPANAuthors');

  my $dist    = $packages->distribution('I/IS/ISHIGAKI/Acme-CPANAuthors-0.12.tar.gz');

  my $latest  = $packages->latest_distribution('Acme-CPANAuthors');

=head1 DESCRIPTION

This is a subset of L<Parse::CPAN::Packages>. The reading
methods are similar in general (accessors are marked as
read-only, though). Internals and data-parsing methods may
be different, but you usually don't need to care.

=head1 METHODS

=head2 new

always takes a file name (both raw C<.txt> file and C<.txt.gz>
file name are acceptable). Raw content of the file is not
acceptable.

=head2 package

takes a name of a package, and returns an object that represents it.

=head2 distribution

takes a file name of a distribution, and returns an object that
represents it.

=head2 latest_distribution

takes a name of a distribution, and returns an object that represents
the latest version of it.

=head2 packages, distributions, latest_distributions

returns a list of stored packages or (latest) distribution objects.

=head2 package_count, distribution_count, latest_distribution_count

returns the number of stored packages or (latest) distributions.

=head1 PREAMBLE ACCESSORS

=head2 file, url, description, columns, intended_for, written_by, line_count, last_updated

These are accessors to the preamble information of
C<02packages.details.txt>.

=head1 PACKAGE ACCESSORS

=head2 package (name), version

  my $package = $packages->package('Acme::CPANAuthors');
  print $package->package, "\n"; # Acme::CPANAuthors
  print $package->version, "\n"; # 0.12

=head2 distribution

returns an object that represents the distribution that the package
belongs to.

=head1 DISTRIBUTION ACCESSORS

=head2 prefix (path), dist (name), version, maturity, filename, cpanid (pauseid), distvname

  my $dist = $packages->distribution('I/IS/ISHIGAKI/Acme-CPANAuthors-0.12.tar.gz');
  print $dist->prefix,"\n";    # I/IS/ISHIGAKI/Acme-CPANAuthors-0.12.tar.gz
  print $dist->dist,"\n";      # Acme-CPANAuthors
  print $dist->version,"\n";   # 0.12
  print $dist->maturity,"\n";  # released
  print $dist->filename,"\n";  # Acme-CPANAuthors-0.12.tar.gz
  print $dist->cpanid,"\n";    # ISHIGAKI
  print $dist->distvname,"\n"; # Acme-CPANAuthors-0.12

=head2 packages (contains)

returns a list of objects that represent the packages the
distribution contains. C<packages> method returns it as an
array reference, C<contains> returns it as an array.

=head1 SEE ALSO

L<Parse::CPAN::Packages>

=head1 AUTHOR

Kenichi Ishigaki, E<lt>ishigaki@cpan.orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2010 by Kenichi Ishigaki.

This program is free software; you can redistribute it and/or
modify it under the same terms as Perl itself.

=cut


lib/Acme/CPANAuthors/Utils/CPANIndex.pm
package Acme::CPANAuthors::Utils::CPANIndex;

use strict;
use warnings;
use Carp;

sub new {
  my $class = shift;
  my $self  = bless { preambles => {} }, $class;

  $self->_install_methods;
  $self->{$_} = {} for keys %{ $self->_mappings };

  $self->_parse(@_) if @_;

  $self;
}

sub _mappings  {+{}}
sub _preambles {}

sub _install_methods {
  my $self = shift;
  my $class = ref $self;

  no strict 'refs';
  no warnings 'redefine';
  for my $method ($self->_preambles) {
    *{"$class\::$method"} = sub {
      my $self = shift;
      $self->{preambles}{$method};
    }
  }

  for my $method (keys %{ $self->_mappings }) {
    my $key = $self->_mappings->{$method};
    *{"$class\::$method"} = sub {
      my ($self, $name) = @_;
      $self->{$key}{$name};
    };
    *{"$class\::${method}s"} = sub {
      my $self = shift;
      values %{ $self->{$key} };
    };
    *{"$class\::${method}_count"} = sub {
      my $self = shift;
      scalar values %{ $self->{$key} };
    };
  }
}

sub _handle {
  my ($self, $file) = @_;

  my $handle;
  if ($file =~ /\.gz$/) {
    require IO::Uncompress::Gunzip;
    $handle = IO::Uncompress::Gunzip->new($file) or croak "Failed to read $file";
  }
  else {
    require IO::File;
    $handle = IO::File->new($file, 'r') or croak "Failed to read $file";
    binmode $handle;
  }
  $handle;
}

1;

__END__

=head1 NAME

Acme::CPANAuthors::Utils::CPANIndex

=head1 SYNOPSIS

=head1 DESCRIPTION

=head1 METHODS

=head2 new

=head1 AUTHOR

Kenichi Ishigaki, E<lt>ishigaki@cpan.orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2010 by Kenichi Ishigaki.

This program is free software; you can redistribute it and/or
modify it under the same terms as Perl itself.

=cut


lib/Acme/CPANAuthors.pm
package Acme::CPANAuthors;

use strict;
use warnings;
use Carp;
use Acme::CPANAuthors::Utils qw( cpan_authors cpan_packages );

our $VERSION = '0.16';

sub new {
  my ($class, @categories) = @_;

  @categories = _list_categories() unless @categories;

  my %authors;
  foreach my $category ( @categories ) {
    %authors = ( %authors, _get_authors_of($category) );
  }
  bless \%authors, $class;
}

sub count {
  my $self = shift;

  return scalar keys %{ $self };
}

sub id {
  my ($self, $id) = @_;

  unless ( $id ) {
    return sort keys %{ $self };
  }
  else {
    return $self->{$id} ? 1 : 0;
  }
}

sub name {
  my ($self, $id) = @_;

  unless ( $id ) {
    return sort values %{ $self };
  }
  else {
    return $self->{$id};
  }
}

sub distributions {
  my ($self, $id) = @_;

  return unless $id;

  my @packages;
  foreach my $package ( cpan_packages->distributions ) {
    if ( $package->cpanid eq $id ) {
      push @packages, $package;
    }
  }

  return @packages;
}

sub latest_distributions {
  my ($self, $id) = @_;

  return unless $id;

  my @packages;
  foreach my $package ( cpan_packages->latest_distributions ) {
    if ( $package->cpanid eq $id ) {
      push @packages, $package;
    }
  }

  return @packages;
}

sub avatar_url {
  my ($self, $id, %options) = @_;

  return unless $id;

  require Gravatar::URL;
  my $author = cpan_authors->author($id) or return;

  return Gravatar::URL::gravatar_url( email => $author->email, %options );
}

sub kwalitee {
  my ($self, $id) = @_;

  return unless $id;

  require Acme::CPANAuthors::Utils::Kwalitee;
  return  Acme::CPANAuthors::Utils::Kwalitee->fetch($id);
}

sub look_for {
  my ($self, $id_or_name) = @_;

  return unless defined $id_or_name;
  unless (ref $id_or_name eq 'Regexp') {
    $id_or_name = qr/$id_or_name/i;
  }

  my @found;
  foreach my $category ( _list_categories() ) {
    my %authors = _get_authors_of($category);
    while ( my ($id, $name) = each %authors ) {
      if ($id =~ /$id_or_name/ or $name =~ /$id_or_name/) {
        push @found, {
          id       => $id,
          name     => $name,
          category => $category,
        };
      }
    }
  }
  return @found;
}

sub _list_categories {
  require Module::Find;
  return grep { $_ !~ /^(?:Register|Utils|Not|Search)$/ }
         map  { s/^Acme::CPANAuthors:://; $_ }
         Module::Find::findsubmod( 'Acme::CPANAuthors' );
}

sub _get_authors_of {
  my $category = shift;

  $category =~ s/^Acme::CPANAuthors:://;

  return if $category =~ /^(?:Register|Utils|Search)$/;

  my $package = "Acme::CPANAuthors\::$category";
  eval "require $package";
  if ( $@ ) {
    carp "$category CPAN Authors are not registered yet: $@";
    return;
  }
  $package->authors;
}

1;

__END__

=head1 NAME

Acme::CPANAuthors - We are CPAN authors

=head1 SYNOPSIS

    use Acme::CPANAuthors;

    my $authors = Acme::CPANAuthors->new('Japanese');

    my $number   = $authors->count;
    my @ids      = $authors->id;
    my @distros  = $authors->distributions('ISHIGAKI');
    my $url      = $authors->avatar_url('ISHIGAKI');
    my $kwalitee = $authors->kwalitee('ISHIGAKI');
    my @info     = $authors->look_for('ishigaki');

  If you don't like this interface, just use a specific authors list.

    use Acme::CPANAuthors::Japanese;

    my %authors = Acme::CPANAuthors::Japanese->authors;

    # note that ->author is context sensitive.
    # however, you can't write this without dereference
    # as "keys" checks the type (actually, the number) of args.
    for my $name (keys %{ Acme::CPANAuthors::Japanese->authors }) {
      print Acme::CPANAuthors::Japanese->authors->{$name}, "\n";
    }

=head1 DESCRIPTION

Sometimes we just want to know something to confirm we're not
alone, or to see if we're doing right things, or to look for
someone we can rely on. This module provides you some basic
information on us.

=head1 WHY THIS MODULE?

We've been holding a Kwalitee competition for Japanese CPAN Authors
since 2006. Though Japanese names are rather easy to distinguish
from Westerner's names (as our names have lots of vowels), it's
tedious to look for Japanese authors every time we hold the contest.
That's why I wrote this module and started maintaining the Japanese
authors list with a script to look for candidates whose name looks
like Japanese by the help of L<Lingua::JA::Romaji::Valid> I coined.

Since then, dozens of lists are uploaded on CPAN. It may be time
to start other games, like offering more useful statistics online.

=head1 ENVIRONMENTAL VARIABLE

=head2 ACME_CPANAUTHORS_HOME

Since 0.14, Acme::CPANAuthors checks C<ACME_CPANAUTHORS_HOME>
environmental variable to look for a place where CPAN indices
are located. If you have a local (mini) CPAN mirror, or a source
directory for your CPAN clients (C<~/.cpan/sources> etc), set
the variable to point there. If not specified, the indices will
be downloaded from the CPAN (to your temporary directory, or
to the current directory).

=head1 METHODS

=head2 new

creates an object and loads the subclasses you specified.
If you don't specify any subclasses, it tries to load all
the subclasses found just under the "Acme::CPANAuthors"
namespace (except L<Acme::CPANAuthors::Not>).

=head2 count

returns how many CPAN authors are registered.

=head2 id

returns all the registered ids by default. If called with an
id, this returns if there's a registered author of the id.

=head2 name

returns all the registered authors' name by default. If called
with an id, this returns the name of the author of the id.

=head2 distributions, latest_distributions

returns an array of Parse::CPAN::Packages::Distribution objects
for the author of the id. See L<Parse::CPAN::Packages> for details.

=head2 avatar_url

returns gravatar url of the id shown at search.cpan.org.
see L<http://site.gravatar.com/site/implement> for details.

=head2 kwalitee

returns kwalitee information for the author of the id.
This information is scraped from http://kwalitee.perl.org/.

=head2 look_for

  my @authors = Acme::CPANAuthors->look_for('SOMEONE');
  foreach my $author (@authors) {
    printf "%s (%s) belongs to %s.\n",
      $author->{id}, $author->{name}, $author->{category};
  }

takes an id or a name (or a part of them, or even a regexp)
and returns an array of hash references, each of which contains
an id, a name, and a basename of the class where the person is
registered. Note that this will load all the installed
Acme::CPANAuthors:: modules but L<Acme::CPANAuthors::Not> and
modules with deeper namespaces.

=head1 SEE ALSO

As of writing this, there're more than a dozen of lists on the CPAN,
including:

=over 4

=item L<Acme::CPANAuthors::Arabic>

=item L<Acme::CPANAuthors::Austrian>

=item L<Acme::CPANAuthors::Brazilian>

=item L<Acme::CPANAuthors::Canadian>

=item L<Acme::CPANAuthors::Chinese>

=item L<Acme::CPANAuthors::Dutch>

=item L<Acme::CPANAuthors::French>

=item L<Acme::CPANAuthors::German>

=item L<Acme::CPANAuthors::Icelandic>

=item L<Acme::CPANAuthors::Indonesian>

=item L<Acme::CPANAuthors::Israeli>

=item L<Acme::CPANAuthors::Italian>

=item L<Acme::CPANAuthors::Japanese>

=item L<Acme::CPANAuthors::Korean>

=item L<Acme::CPANAuthors::Norwegian>

=item L<Acme::CPANAuthors::Portuguese>

=item L<Acme::CPANAuthors::Russian>

=item L<Acme::CPANAuthors::Taiwanese>

=item L<Acme::CPANAuthors::Turkish>

=item L<Acme::CPANAuthors::Ukrainian>

=back

These are not regional ones but for some local groups.

=over 4

=item L<Acme::CPANAuthors::Booking>

=item L<Acme::CPANAuthors::British::Companies>

=item L<Acme::CPANAuthors::CodeRepos>

=item L<Acme::CPANAuthors::GeekHouse>

=item L<Acme::CPANAuthors::GitHub>

=back

These are lists for specific module authors.

=over 4

=item L<Acme::CPANAuthors::AnyEvent>

=item L<Acme::CPANAuthors::DualLife>

=item L<Acme::CPANAuthors::POE>

=item L<Acme::CPANAuthors::ToBeLike>

=item L<Acme::CPANAuthors::Acme::CPANAuthors::Authors>

=back

And other stuff.

=over 4

=item L<Acme::CPANAuthors::CPANTS::FiveOrMore>

=item L<Acme::CPANAuthors::Female>

=item L<Acme::CPANAuthors::Misanthrope>

=item L<Acme::CPANAuthors::Not>

=item L<Acme::CPANAuthors::Pumpkings>

=item L<Acme::CPANAuthors::Search> (now L<CPAN::AuthorsSearch>)

=item L<Acme::CPANAuthors::You::re_using>

=back

Thank you all. And I hope more to come.

=head1 AUTHOR

Kenichi Ishigaki, E<lt>ishigaki at cpan.orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2007-2010 by Kenichi Ishigaki.

This program is free software; you can redistribute it and/or
modify it under the same terms as Perl itself.

=cut
